src/core/public/plugins/plugins_service.ts

```diff
@@ -98,34 +96,10 @@ export class PluginsService implements CoreService<PluginsServiceSetup, PluginsS
         {} as Record<PluginName, unknown>
       );
 
-      let contract: unknown;
-      const contractOrPromise = plugin.setup(
+      const contract = plugin.setup(
         createPluginSetupContext(this.coreContext, deps, plugin),
         pluginDepContracts
       );
-      if (isPromise(contractOrPromise)) {
-        if (this.coreContext.env.mode.dev) {
-          // eslint-disable-next-line no-console
-          console.log(
-            `Plugin ${pluginName} is using asynchronous setup lifecycle. Asynchronous plugins support will be removed in a later version.`
-          );
-        }
```

Comment:
I just wonder if we shouldn't keep this block and now throw instead of just logging a warning? It would avoid developer mistakes if a new plugin is defining async `setup` or `start` functions by mistake?

Reply:
Sorry! I had auto-merge enabled. I can create a follow-up PR to add this. Do you think it's worth it?

Reply:
As discussed on slack, it would be a nice to have, just to get a more explicit error during the dev process, but it's really not strictly necessary.

Reply:
We discussed it on Slack: if a plugin exposes a _promised_ contract and declares it as `class MyPlugin extends Plugin<Promise<Contract>>`, it's technically correct and we shouldn't throw.

The plugins using that contract will have to await the contract, which doesn't make it a good contract according to Kibana's architecture, but that's a different discussion.

If we see it's really needed in the future, we can add the validation later on.